### PR TITLE
[Kernel] Add Fused Layernorm + Dynamic-Per-Token Quant Kernels

### PR DESCRIPTION
**Pull request by varun-sundar-rabindranath**
Add Fused Layernorm + Dynamic-Per-Token Quant kernels.

Numbers:
Take away: The fused kernels out perform the unfused counterparts. It is important to note that we don't yet know how much performance gain is actually coming from the fusion (the fused kernels have vectorization optimizations that are not necessarily in their unfused counterparts).  

GPU : A6000 x 1
Command: `python3 benchmarks/fused_kernels/layernorm_rms_benchmarks.py`

<google-sheets-html-origin><style type="text/css"></style>
num-tokens x hidden-size x add-residual x input-dtype | unfused_int8_impl (us) | unfused_fp8_impl (us) | fused_int8_impl (us) | fused_fp8_impl (us)
-- | -- | -- | -- | --
N 1 x D 1024 x R True x DT torch.bfloat16 | 22.1 | 25.4 | 17.3 | 17.3
N 1 x D 1024 x R True x DT torch.float32 | 22.5 | 25.5 | 17 | 16.7
N 1 x D 1024 x R False x DT torch.bfloat16 | 25 | 28 | 16.8 | 16.9
N 1 x D 1024 x R False x DT torch.float32 | 24.8 | 27.8 | 17.1 | 17.1
N 1 x D 2048 x R True x DT torch.bfloat16 | 22.5 | 25.6 | 17.1 | 17.4
N 1 x D 2048 x R True x DT torch.float32 | 22.5 | 25.6 | 17.1 | 17.2
N 1 x D 2048 x R False x DT torch.bfloat16 | 24.7 | 27.8 | 17.2 | 17
N 1 x D 2048 x R False x DT torch.float32 | 25.1 | 28 | 16.8 | 16.9
N 1 x D 3072 x R True x DT torch.bfloat16 | 22.4 | 25.4 | 17.2 | 17.3
N 1 x D 3072 x R True x DT torch.float32 | 22.3 | 25.5 | 16.9 | 17.2
N 1 x D 3072 x R False x DT torch.bfloat16 | 25 | 27.9 | 17 | 17.2
N 1 x D 3072 x R False x DT torch.float32 | 25.1 | 27.9 | 17 | 17
N 1 x D 4096 x R True x DT torch.bfloat16 | 22.5 | 25.5 | 17 | 17.4
N 1 x D 4096 x R True x DT torch.float32 | 22.4 | 25.7 | 17.3 | 17.3
N 1 x D 4096 x R False x DT torch.bfloat16 | 25.2 | 28.1 | 16.9 | 17.1
N 1 x D 4096 x R False x DT torch.float32 | 24.9 | 28.2 | 17 | 17.2
N 1 x D 5120 x R True x DT torch.bfloat16 | 22.5 | 25.5 | 17.3 | 17.4
N 1 x D 5120 x R True x DT torch.float32 | 22.5 | 25.6 | 17.2 | 17.1
N 1 x D 5120 x R False x DT torch.bfloat16 | 25 | 28.1 | 17.3 | 17.3
N 1 x D 5120 x R False x DT torch.float32 | 25.2 | 28.2 | 17.2 | 17.4
N 1 x D 6144 x R True x DT torch.bfloat16 | 22.4 | 25.7 | 17.4 | 17.3
N 1 x D 6144 x R True x DT torch.float32 | 22.7 | 25.8 | 17.2 | 17.3
N 1 x D 6144 x R False x DT torch.bfloat16 | 25.3 | 28.1 | 17.1 | 17.4
N 1 x D 6144 x R False x DT torch.float32 | 26.5 | 28.1 | 16.4 | 16.5
N 1 x D 7168 x R True x DT torch.bfloat16 | 22.6 | 25.6 | 16.5 | 16.5
N 1 x D 7168 x R True x DT torch.float32 | 22.6 | 25.4 | 16.5 | 16.5
N 1 x D 7168 x R False x DT torch.bfloat16 | 25.2 | 27.9 | 16.5 | 16.6
N 1 x D 7168 x R False x DT torch.float32 | 25.2 | 28.2 | 16.5 | 16.7
N 16 x D 1024 x R True x DT torch.bfloat16 | 23.5 | 25.3 | 17 | 17.1
N 16 x D 1024 x R True x DT torch.float32 | 23.7 | 25.3 | 17 | 17.2
N 16 x D 1024 x R False x DT torch.bfloat16 | 26.3 | 27.7 | 17 | 17
N 16 x D 1024 x R False x DT torch.float32 | 26.2 | 27.9 | 17.1 | 17.1
N 16 x D 2048 x R True x DT torch.bfloat16 | 23.5 | 25.6 | 17.1 | 17.1
N 16 x D 2048 x R True x DT torch.float32 | 23.9 | 25.5 | 16.9 | 17.1
N 16 x D 2048 x R False x DT torch.bfloat16 | 26.3 | 27.9 | 17.1 | 17.1
N 16 x D 2048 x R False x DT torch.float32 | 26.5 | 27.7 | 17 | 16.9
N 16 x D 3072 x R True x DT torch.bfloat16 | 23.7 | 25.4 | 17 | 17.1
N 16 x D 3072 x R True x DT torch.float32 | 23.7 | 25.9 | 17.1 | 17.1
N 16 x D 3072 x R False x DT torch.bfloat16 | 26.3 | 28 | 17.1 | 17.1
N 16 x D 3072 x R False x DT torch.float32 | 26.4 | 28 | 16.8 | 17.1
N 16 x D 4096 x R True x DT torch.bfloat16 | 23.6 | 25.2 | 17.1 | 17.1
N 16 x D 4096 x R True x DT torch.float32 | 23.7 | 25.5 | 17 | 17
N 16 x D 4096 x R False x DT torch.bfloat16 | 26.1 | 27.7 | 16.9 | 16.9
N 16 x D 4096 x R False x DT torch.float32 | 26.4 | 28 | 17.1 | 17.1
N 16 x D 5120 x R True x DT torch.bfloat16 | 23.4 | 25.3 | 17.4 | 17.1
N 16 x D 5120 x R True x DT torch.float32 | 23.6 | 25.4 | 17.1 | 17
N 16 x D 5120 x R False x DT torch.bfloat16 | 26.4 | 28.2 | 17.1 | 17
N 16 x D 5120 x R False x DT torch.float32 | 26.3 | 28 | 17.1 | 17.1
N 16 x D 6144 x R True x DT torch.bfloat16 | 23.9 | 25.4 | 17.2 | 17.1
N 16 x D 6144 x R True x DT torch.float32 | 24.2 | 25.4 | 17 | 17.2
N 16 x D 6144 x R False x DT torch.bfloat16 | 26.4 | 27.9 | 17.1 | 17.4
N 16 x D 6144 x R False x DT torch.float32 | 26.6 | 27.9 | 17 | 17.1
N 16 x D 7168 x R True x DT torch.bfloat16 | 23.9 | 25.4 | 17 | 17.3
N 16 x D 7168 x R True x DT torch.float32 | 23.6 | 25.2 | 16.9 | 17.1
N 16 x D 7168 x R False x DT torch.bfloat16 | 26.3 | 27.9 | 17.1 | 17.1
N 16 x D 7168 x R False x DT torch.float32 | 26.4 | 27.9 | 16.9 | 16.9
N 32 x D 1024 x R True x DT torch.bfloat16 | 23.6 | 25.4 | 17 | 17.3
N 32 x D 1024 x R True x DT torch.float32 | 23.5 | 25.6 | 17.3 | 17.2
N 32 x D 1024 x R False x DT torch.bfloat16 | 26.5 | 28.1 | 17.1 | 17.2
N 32 x D 1024 x R False x DT torch.float32 | 26.5 | 28.1 | 17.1 | 17.3
N 32 x D 2048 x R True x DT torch.bfloat16 | 23.9 | 25.4 | 17.3 | 17.4
N 32 x D 2048 x R True x DT torch.float32 | 23.7 | 25.4 | 17.2 | 17.3
N 32 x D 2048 x R False x DT torch.bfloat16 | 26.5 | 28 | 17.3 | 17.2
N 32 x D 2048 x R False x DT torch.float32 | 26.4 | 28.2 | 17.1 | 17.2
N 32 x D 3072 x R True x DT torch.bfloat16 | 24.1 | 25.6 | 17.4 | 17.4
N 32 x D 3072 x R True x DT torch.float32 | 24.1 | 25.5 | 17.3 | 17.2
N 32 x D 3072 x R False x DT torch.bfloat16 | 26.5 | 28.1 | 17.2 | 17.2
N 32 x D 3072 x R False x DT torch.float32 | 26.5 | 27.8 | 17.3 | 17.2
N 32 x D 4096 x R True x DT torch.bfloat16 | 24 | 25.5 | 17 | 17.4
N 32 x D 4096 x R True x DT torch.float32 | 23.8 | 25.5 | 17.1 | 17.2
N 32 x D 4096 x R False x DT torch.bfloat16 | 26.4 | 27.8 | 16.9 | 17.1
N 32 x D 4096 x R False x DT torch.float32 | 26.5 | 28.1 | 16.9 | 17.1
N 32 x D 5120 x R True x DT torch.bfloat16 | 23.9 | 25.4 | 17.1 | 17.2
N 32 x D 5120 x R True x DT torch.float32 | 23.6 | 25.5 | 17.1 | 17
N 32 x D 5120 x R False x DT torch.bfloat16 | 26.3 | 28 | 17.1 | 17
N 32 x D 5120 x R False x DT torch.float32 | 26.4 | 28.2 | 17.2 | 17.1
N 32 x D 6144 x R True x DT torch.bfloat16 | 23.7 | 25.3 | 17.1 | 17.3
N 32 x D 6144 x R True x DT torch.float32 | 23.9 | 25.4 | 17 | 17.2
N 32 x D 6144 x R False x DT torch.bfloat16 | 26.6 | 28.2 | 17 | 17
N 32 x D 6144 x R False x DT torch.float32 | 26.5 | 28 | 17.2 | 17.2
N 32 x D 7168 x R True x DT torch.bfloat16 | 23.9 | 25.5 | 17.1 | 17.2
N 32 x D 7168 x R True x DT torch.float32 | 23.8 | 25.4 | 17 | 17.5
N 32 x D 7168 x R False x DT torch.bfloat16 | 26.4 | 28.2 | 17 | 17.3
N 32 x D 7168 x R False x DT torch.float32 | 26.3 | 28.1 | 17.1 | 17.2
N 64 x D 1024 x R True x DT torch.bfloat16 | 23.9 | 25.5 | 17.2 | 17.3
N 64 x D 1024 x R True x DT torch.float32 | 24 | 25.4 | 17.1 | 17.2
N 64 x D 1024 x R False x DT torch.bfloat16 | 26.4 | 27.9 | 17.1 | 17.2
N 64 x D 1024 x R False x DT torch.float32 | 26.5 | 28 | 17 | 17.1
N 64 x D 2048 x R True x DT torch.bfloat16 | 24.1 | 25.6 | 17.3 | 17.2
N 64 x D 2048 x R True x DT torch.float32 | 24 | 25.7 | 17.3 | 17.2
N 64 x D 2048 x R False x DT torch.bfloat16 | 26.4 | 28.1 | 17 | 17.2
N 64 x D 2048 x R False x DT torch.float32 | 26 | 28.1 | 17.1 | 17.3
N 64 x D 3072 x R True x DT torch.bfloat16 | 24 | 25.7 | 17.2 | 17.4
N 64 x D 3072 x R True x DT torch.float32 | 23.5 | 25.5 | 17.1 | 17.4
N 64 x D 3072 x R False x DT torch.bfloat16 | 26.4 | 27.9 | 17.2 | 17.2
N 64 x D 3072 x R False x DT torch.float32 | 26.7 | 28.1 | 17.2 | 17
N 64 x D 4096 x R True x DT torch.bfloat16 | 23.9 | 25.4 | 17.1 | 17.3
N 64 x D 4096 x R True x DT torch.float32 | 23.7 | 25.2 | 17.2 | 17.4
N 64 x D 4096 x R False x DT torch.bfloat16 | 26.4 | 27.9 | 17.2 | 17.2
N 64 x D 4096 x R False x DT torch.float32 | 26.4 | 28.2 | 17 | 17.2
N 64 x D 5120 x R True x DT torch.bfloat16 | 24 | 25.5 | 17.1 | 17.4
N 64 x D 5120 x R True x DT torch.float32 | 23.9 | 25.4 | 16.9 | 17
N 64 x D 5120 x R False x DT torch.bfloat16 | 26.3 | 27.9 | 17 | 16.9
N 64 x D 5120 x R False x DT torch.float32 | 26.2 | 27.8 | 17 | 17.1
N 64 x D 6144 x R True x DT torch.bfloat16 | 23.9 | 25.6 | 17 | 17.1
N 64 x D 6144 x R True x DT torch.float32 | 23.7 | 25.7 | 16.9 | 16.9
N 64 x D 6144 x R False x DT torch.bfloat16 | 26 | 28.1 | 17 | 17
N 64 x D 6144 x R False x DT torch.float32 | 25.9 | 28 | 16.9 | 16.9
N 64 x D 7168 x R True x DT torch.bfloat16 | 23.5 | 25.5 | 17 | 16.9
N 64 x D 7168 x R True x DT torch.float32 | 23.7 | 25.5 | 16.9 | 16.9
N 64 x D 7168 x R False x DT torch.bfloat16 | 26.1 | 27.9 | 17 | 17.3
N 64 x D 7168 x R False x DT torch.float32 | 26.6 | 28.1 | 17.1 | 17.4
N 128 x D 1024 x R True x DT torch.bfloat16 | 24 | 25.4 | 16.9 | 17.3
N 128 x D 1024 x R True x DT torch.float32 | 24.1 | 25.4 | 16.9 | 17.1
N 128 x D 1024 x R False x DT torch.bfloat16 | 26.1 | 28 | 16.8 | 17.1
N 128 x D 1024 x R False x DT torch.float32 | 26.2 | 27.9 | 16.9 | 17
N 128 x D 2048 x R True x DT torch.bfloat16 | 23.8 | 25.7 | 17 | 17.2
N 128 x D 2048 x R True x DT torch.float32 | 23.7 | 25.5 | 17.1 | 17.1
N 128 x D 2048 x R False x DT torch.bfloat16 | 26.4 | 28.1 | 17.2 | 17.1
N 128 x D 2048 x R False x DT torch.float32 | 26 | 27.9 | 16.9 | 17.3
N 128 x D 3072 x R True x DT torch.bfloat16 | 23.5 | 25.5 | 17 | 17
N 128 x D 3072 x R True x DT torch.float32 | 23.6 | 25.6 | 17.1 | 17.2
N 128 x D 3072 x R False x DT torch.bfloat16 | 26.3 | 28 | 17.2 | 17.2
N 128 x D 3072 x R False x DT torch.float32 | 26.3 | 27.9 | 17.2 | 17.2
N 128 x D 4096 x R True x DT torch.bfloat16 | 23.8 | 25.5 | 17.1 | 17.1
N 128 x D 4096 x R True x DT torch.float32 | 23.9 | 25.6 | 17 | 17.2
N 128 x D 4096 x R False x DT torch.bfloat16 | 26.4 | 27.8 | 17.3 | 17.4
N 128 x D 4096 x R False x DT torch.float32 | 26.4 | 27.9 | 17.2 | 17.3
N 128 x D 5120 x R True x DT torch.bfloat16 | 23.6 | 25.6 | 17.2 | 17.1
N 128 x D 5120 x R True x DT torch.float32 | 24.1 | 25.7 | 17.2 | 17.3
N 128 x D 5120 x R False x DT torch.bfloat16 | 26.3 | 27.9 | 17.1 | 17.1
N 128 x D 5120 x R False x DT torch.float32 | 26.3 | 28 | 16.9 | 17.1
N 128 x D 6144 x R True x DT torch.bfloat16 | 24.1 | 25.5 | 17.2 | 17.2
N 128 x D 6144 x R True x DT torch.float32 | 24.1 | 28.5 | 17.2 | 17.3
N 128 x D 6144 x R False x DT torch.bfloat16 | 26.2 | 28 | 17.3 | 17.3
N 128 x D 6144 x R False x DT torch.float32 | 26.4 | 28.1 | 17.1 | 17.2
N 128 x D 7168 x R True x DT torch.bfloat16 | 24.1 | 25.9 | 17.3 | 17.5
N 128 x D 7168 x R True x DT torch.float32 | 27.8 | 34.8 | 20.3 | 21.4
N 128 x D 7168 x R False x DT torch.bfloat16 | 26.5 | 28.2 | 16.9 | 17.1
N 128 x D 7168 x R False x DT torch.float32 | 26.2 | 28.6 | 17.1 | 16.9
N 256 x D 1024 x R True x DT torch.bfloat16 | 23.6 | 25.6 | 17 | 17.1
N 256 x D 1024 x R True x DT torch.float32 | 23.6 | 25.7 | 17 | 17.2
N 256 x D 1024 x R False x DT torch.bfloat16 | 26.4 | 28.2 | 17 | 17.2
N 256 x D 1024 x R False x DT torch.float32 | 26.5 | 28 | 17 | 17
N 256 x D 2048 x R True x DT torch.bfloat16 | 23.6 | 25.2 | 17 | 17
N 256 x D 2048 x R True x DT torch.float32 | 23.7 | 25.6 | 17.1 | 17.2
N 256 x D 2048 x R False x DT torch.bfloat16 | 26.5 | 28.2 | 17 | 16.9
N 256 x D 2048 x R False x DT torch.float32 | 26.4 | 28.1 | 17 | 17.2
N 256 x D 3072 x R True x DT torch.bfloat16 | 24.3 | 25.9 | 17.3 | 17.3
N 256 x D 3072 x R True x DT torch.float32 | 24.3 | 31.2 | 17.3 | 17.3
N 256 x D 3072 x R False x DT torch.bfloat16 | 26.3 | 28.2 | 17.2 | 17.1
N 256 x D 3072 x R False x DT torch.float32 | 26.7 | 28.4 | 17.2 | 17.2
N 256 x D 4096 x R True x DT torch.bfloat16 | 23.7 | 25.6 | 17.4 | 17.5
N 256 x D 4096 x R True x DT torch.float32 | 32.6 | 41.9 | 23 | 24.4
N 256 x D 4096 x R False x DT torch.bfloat16 | 26.5 | 28.2 | 17.3 | 17.3
N 256 x D 4096 x R False x DT torch.float32 | 26.9 | 36 | 17.1 | 17.1
N 256 x D 5120 x R True x DT torch.bfloat16 | 23.8 | 27 | 16.8 | 18
N 256 x D 5120 x R True x DT torch.float32 | 45.4 | 53.6 | 32.3 | 35.8
N 256 x D 5120 x R False x DT torch.bfloat16 | 26.5 | 30.2 | 17.3 | 17.2
N 256 x D 5120 x R False x DT torch.float32 | 32.9 | 41.8 | 17.2 | 17
N 256 x D 6144 x R True x DT torch.bfloat16 | 25.7 | 34.5 | 20.3 | 26.4
N 256 x D 6144 x R True x DT torch.float32 | 58.4 | 67.9 | 36.8 | 39.2
N 256 x D 6144 x R False x DT torch.bfloat16 | 28.5 | 36.8 | 17 | 17
N 256 x D 6144 x R False x DT torch.float32 | 41 | 50.2 | 18.2 | 20.5
N 256 x D 7168 x R True x DT torch.bfloat16 | 31 | 39.7 | 23.7 | 27.4
N 256 x D 7168 x R True x DT torch.float32 | 69.7 | 80.9 | 42.7 | 44.7
N 256 x D 7168 x R False x DT torch.bfloat16 | 32.2 | 41.4 | 17.2 | 17.1
N 256 x D 7168 x R False x DT torch.float32 | 48.9 | 59.6 | 19.7 | 22.2
N 512 x D 1024 x R True x DT torch.bfloat16 | 23.5 | 25.7 | 17.4 | 17.1
N 512 x D 1024 x R True x DT torch.float32 | 23.8 | 25.5 | 17.1 | 17.2
N 512 x D 1024 x R False x DT torch.bfloat16 | 26.7 | 28.1 | 17.2 | 17
N 512 x D 1024 x R False x DT torch.float32 | 26.6 | 28.5 | 17 | 17.4
N 512 x D 2048 x R True x DT torch.bfloat16 | 23.9 | 26.6 | 17.3 | 17.4
N 512 x D 2048 x R True x DT torch.float32 | 37.2 | 49.4 | 24.6 | 25.2
N 512 x D 2048 x R False x DT torch.bfloat16 | 26.4 | 31.9 | 17.1 | 17.3
N 512 x D 2048 x R False x DT torch.float32 | 29 | 40.8 | 17.2 | 17.2
N 512 x D 3072 x R True x DT torch.bfloat16 | 29.3 | 41 | 21.7 | 24.5
N 512 x D 3072 x R True x DT torch.float32 | 59.6 | 73.7 | 35.4 | 35.5
N 512 x D 3072 x R False x DT torch.bfloat16 | 30.5 | 41.9 | 17.2 | 18.1
N 512 x D 3072 x R False x DT torch.float32 | 40.2 | 51.9 | 18.9 | 21
N 512 x D 4096 x R True x DT torch.bfloat16 | 40 | 52.3 | 26.4 | 29.1
N 512 x D 4096 x R True x DT torch.float32 | 82.9 | 99.4 | 45.8 | 46.1
N 512 x D 4096 x R False x DT torch.bfloat16 | 40.2 | 50.7 | 18.4 | 21.6
N 512 x D 4096 x R False x DT torch.float32 | 52.9 | 69.2 | 21.7 | 24.5
N 512 x D 5120 x R True x DT torch.bfloat16 | 53.6 | 64.3 | 36 | 43.4
N 512 x D 5120 x R True x DT torch.float32 | 106 | 126.5 | 57.2 | 64.7
N 512 x D 5120 x R False x DT torch.bfloat16 | 49.3 | 59.4 | 24.1 | 29
N 512 x D 5120 x R False x DT torch.float32 | 65.3 | 83.8 | 28.5 | 32.8
N 512 x D 6144 x R True x DT torch.bfloat16 | 65.7 | 78.2 | 38.8 | 44.9
N 512 x D 6144 x R True x DT torch.float32 | 127.4 | 151 | 67.3 | 71.8
N 512 x D 6144 x R False x DT torch.bfloat16 | 57.6 | 68.8 | 25.7 | 31.1
N 512 x D 6144 x R False x DT torch.float32 | 75.3 | 98.5 | 30.4 | 34.7
N 512 x D 7168 x R True x DT torch.bfloat16 | 79.7 | 93.3 | 43.6 | 48.9
N 512 x D 7168 x R True x DT torch.float32 | 149.5 | 176.6 | 77.5 | 80.4
N 512 x D 7168 x R False x DT torch.bfloat16 | 66 | 79.1 | 27.4 | 33.3
N 512 x D 7168 x R False x DT torch.float32 | 88.2 | 113.7 | 33.6 | 37.5

 
---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by varun-sundar-rabindranath** [comment]
Marked it draft for now. Will open it up after some performance profiling.

File: csrc/quantization/fused_kernels/layernorm_utils.cuh [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+#include "quant_conversions.cuh"
+
+namespace vllm {
+
+// has_residual must be true, if residual is not a nullptr
+template <typename scalar_t, bool has_residual = false>
+__device__ void compute_rms(float* rms, scalar_t const* __restrict__ input,
+                            int const hidden_size, float const epsilon,
+                            scalar_t const* __restrict__ residual = nullptr) {
+  int const token_offset = blockIdx.x * hidden_size;
```
Comment (tlrmchlsmth) [inline csrc/quantization/fused_kernels/layernorm_utils.cuh:17]: this should be an int64_t

File: csrc/quantization/fused_kernels/layernorm_utils.cuh [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+}
+
+template <typename scalar_t, typename scalar_out_t, bool has_residual = false>
+__device__ void compute_dynamic_per_token_scales(
+    float* __restrict__ token_scale, float* __restrict__ all_token_scales,
+    scalar_t const* __restrict__ input, scalar_t const* __restrict__ weight,
+    float const rms, float const* __restrict__ scale_ub,
+    float const min_scaling_factor, int const hidden_size,
+    scalar_t const* __restrict__ residual = nullptr) {
+  int const token_offset = blockIdx.x * hidden_size;
```
Comment (tlrmchlsmth) [inline csrc/quantization/fused_kernels/layernorm_utils.cuh:46]: ditto, make this int64_t

File: csrc/quantization/fused_kernels/layernorm_utils.cuh [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+
+template <typename scalar_t, typename scalar_out_t, bool is_scale_inverted,
+          bool has_residual = false>
+__device__ void norm_and_quant(scalar_out_t* __restrict__ output,
+                               scalar_t const* __restrict__ input,
+                               scalar_t const* __restrict__ weight,
+                               float const rms, float const scale,
+                               int const hidden_size,
+                               scalar_t* __restrict__ residual = nullptr) {
+  int const token_offset = blockIdx.x * hidden_size;
```
Comment (tlrmchlsmth) [inline csrc/quantization/fused_kernels/layernorm_utils.cuh:87]: ditto: int64_t

File: csrc/quantization/fused_kernels/layernorm_utils.cuh [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+
+namespace vectorized {
+
+// Compute 1.0/rms(input)
+// hidden_size must be a multiple of 4
+template <typename scalar_t, bool has_residual = false>
+__device__ void compute_rms(float* rms, scalar_t const* __restrict__ input,
+                            int const hidden_size, float const epsilon,
+                            scalar_t const* __restrict__ residual = nullptr) {
+  int const token_offset = blockIdx.x * hidden_size;
```
Comment (tlrmchlsmth) [inline csrc/quantization/fused_kernels/layernorm_utils.cuh:111]: ditto, int64_t

File: csrc/quantization/fused_kernels/layernorm_utils.cuh [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+// Vectorized version of vllm::compute_dynamic_per_token_scales
+// hidden_size must be a multiple of 4
+template <typename scalar_t, typename scalar_out_t, bool has_residual = false>
+__device__ void compute_dynamic_per_token_scales(
+    float* __restrict__ token_scale, float* __restrict__ all_token_scales,
+    scalar_t const* __restrict__ input, scalar_t const* __restrict__ weight,
+    float const rms, float const* __restrict__ scale_ub,
+    float const min_scaling_factor, int const hidden_size,
+    scalar_t const* __restrict__ residual = nullptr) {
+  int const token_offset = blockIdx.x * hidden_size;
```
Comment (tlrmchlsmth) [inline csrc/quantization/fused_kernels/layernorm_utils.cuh:170]: ditto, int64_t

File: csrc/quantization/fused_kernels/layernorm_utils.cuh [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+// hidden_size must be a multiple of 4
+template <typename scalar_t, typename scalar_out_t, bool is_scale_inverted,
+          bool has_residual = false>
+__device__ void norm_and_quant(scalar_out_t* __restrict__ output,
+                               scalar_t const* __restrict__ input,
+                               scalar_t const* __restrict__ weight,
+                               float const rms, float const scale,
+                               int const hidden_size,
+                               scalar_t* __restrict__ residual = nullptr) {
+  int const token_offset = blockIdx.x * hidden_size;
```
Comment (tlrmchlsmth) [inline csrc/quantization/fused_kernels/layernorm_utils.cuh:246]: ditto, int64_t

File: csrc/quantization/fused_kernels/quant_conversions.cuh [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+namespace vectorized {
+
+// Vectorized version of scaled_quant_conversion
+template <typename scalar_t, typename quant_type_t, bool is_scale_inverted>
+__device__ void scaled_quant_conversion(quant_type_t* __restrict__ out,
+                                        scalar_t const* __restrict__ input,
+                                        float const scale, int const tid,
+                                        int const num_elems, int const step) {
```
Comment (tlrmchlsmth) [inline csrc/quantization/fused_kernels/quant_conversions.cuh:86]: Are these functions used anywhere?

**Review by tlrmchlsmth — COMMENTED** [review]
Left some in-line comments. I see some potential issues with using int for large activations that will need to be resolved. Along those lines, I thing being explicit about using int32_t instead of int will improve things.

File: csrc/quantization/fused_kernels/quant_conversions.cuh (hunk 7 above)
Comment (varun-sundar-rabindranath) [reply csrc/quantization/fused_kernels/quant_conversions.cuh:86]: No. it isn't yet. I plan to use them in the static-per-tensor case. On second thought, let me remove it from here and introduce it in static-per-tensor PR.

File: csrc/quantization/fused_kernels/layernorm_utils.cuh (hunk 1 above)
Comment (varun-sundar-rabindranath) [reply csrc/quantization/fused_kernels/layernorm_utils.cuh]: Thanks for catching this. Updated the `int`s to `int32_t` / `int64_t` 👍

**Comment by ProExpertProg** [comment]
@varun-sundar-rabindranath close in favor of #10906 ?

**Comment by varun-sundar-rabindranath** [comment]
close in favor of https://github.com/vllm-project/vllm/pull/10906 . Thanks @ProExpertProg

**Pull request closed without merging**